modules/http4s-emitter/src/main/scala/com.snowplowanalytics.snowplow.scalatracker.emitters.http4s/Http4sEmitter.scala

```diff
@@ -0,0 +1,146 @@
+
+  def build[F[_]: Concurrent: Timer](collector: EndpointParams,
+                                     client: Client[F],
+                                     bufferConfig: BufferConfig    = BufferConfig.Default,
+                                     retryPolicy: RetryPolicy      = RetryPolicy.Default,
+                                     queuePolicy: EventQueuePolicy = EventQueuePolicy.Default,
+                                     callback: Option[Callback[F]] = None): Resource[F, Emitter[F]] =
+    Resource {
+      for {
+        queue <- queueForPolicy[F](queuePolicy)
+        fiber <- Concurrent[F].start(drainQueue(queue, client, collector, bufferConfig, retryPolicy, callback))
+      } yield instance(queue) -> shutdown(fiber, queue)
+    }
+
+  private def instance[F[_]](queue: Enqueue[F, Action]): Emitter[F] =
+    new Emitter[F] {
+      override def send(event: Payload): F[Unit] =
+        queue.enqueue1(Action.Enqueue(event))
+
+      override def flushBuffer(): F[Unit] =
+        queue.enqueue1(Action.Flush)
+    }
+
+  private def queueForPolicy[F[_]: Concurrent](policy: EventQueuePolicy): F[Queue[F, Action]] =
+    policy match {
+      case EventQueuePolicy.UnboundedQueue      => Queue.unbounded
+      case EventQueuePolicy.BlockWhenFull(max)  => Queue.bounded(max)
+      case EventQueuePolicy.IgnoreWhenFull(max) => Queue.bounded(max)
+      case EventQueuePolicy.ErrorWhenFull(max)  => Queue.bounded(max)
+    }
+
+  private def shutdown[F[_]: Monad](fiber: Fiber[F, Unit], queue: Enqueue[F, Action]): F[Unit] =
+    queue.enqueue1(Action.Terminate) *> fiber.join
+
+  private def drainQueue[F[_]: Sync: Timer](queue: Dequeue[F, Action],
+                                            client: Client[F],
+                                            collector: EndpointParams,
+                                            bufferConfig: BufferConfig,
+                                            retryPolicy: RetryPolicy,
+                                            callback: Option[Callback[F]]): F[Unit] =
+    Sync[F].delay(new Random()).flatMap { rng =>
+      queue.dequeue
+        .takeThrough(_ != Action.Terminate)
+        .through(bufferEvents(bufferConfig))
+        .evalMap(sendRequest(client, collector, retryPolicy, rng, callback))
+        .compile
+        .drain
+    }
+
+  private def bufferEvents[F[_]](bufferConfig: BufferConfig): Pipe[F, Action, Request] =
+    _.mapAccumulate(Buffer(bufferConfig)) {
+      case (buffer, action) => buffer.handle(action)
+    }.collect {
+      case (_, Some(flushed)) => flushed
+    }
+
+  private def sendRequest[F[_]: Sync: Timer](client: Client[F],
+                                             collector: EndpointParams,
+                                             retryPolicy: RetryPolicy,
+                                             rng: Random,
+                                             callback: Option[Callback[F]]): Request => F[Unit] =
+    Sync[F].tailRecM(_) { request =>
```

Comment:
I honestly never seen this pattern (using `tailRecM` for retries), but looks interesting.